### ISSUE TITLE
observability

### ISSUE THREAD

**Issue by tmbdev** [ipfs/go-ipfs#7079]
The "ipfs" command currently does not seem to have a way of observing what it is doing or what it is waiting for. This makes it difficult to diagnose sources of performance problems and errors in setup. It also seems to lead to lots of questions and submitted issues about "why is this slow" and "what is it doing".

The current "--debug" flag only outputs the config path for me, nothing more.

I'd suggest adding another flag that outputs information about what operations have been started, what operations have been completed, and what operations have timed out. A simple version of such a flag might just print a line for each event, though a screen oriented status display might be nice to have.

**Comment by RubenKelevra** [comment]
ipfs is extremely a mutli-threaded application. It usually doesn't wait for _one_ thing but runs multiple stuff in parallel to do avoid that the application has to wait for network latencies for example.

If you like to see, what's going on, you can just adjust the log level per component. I recommend to enable debugging for all components to see what's going on:

`ipfs log level all debug`

Note that this will decrease the performance, because of the high amount of log output.

**Comment by Stebalien** [comment]
I believe there are actually proposals for this floating around. The idea is that we'd stash some form of eventbus/listener in the request context at the top. Then, we should be able to get back events like "looking for providers", "trying to find block X", "resolving address Y", etc.

**Comment by RubenKelevra** [comment]
@Stebalien you mean like `ipfs log tail` shows some events right now?

**Comment by hsanjuan** [comment]
> I'd suggest adding another flag that outputs information about what operations have been started, what operations have been completed, and what operations have timed out

This sounds a bit like current `ipfs diag cmds` (making sure everyone following this convo knows about, as it may help resolve certain problems like "how long has this API request been hanging").

**Comment by tmbdev** [comment]
Thanks for the suggestions. "ipfs log level all debug" doesn't result in any output per se; "ipfs log tail" shows voluminous JSON output about all activity in the server. "ipfs diag cmds" gives some info about how long commands take as a whole.

Generally speaking, I'm trying to integrate IPFS as a backend store for scientific and machine learning data, but even with ipfs-connecting all the nodes manually, performance is too unpredictable to be useful with the setup I have. That's why I'm looking for better diagnostic tools.

**Comment by ribasushi** [comment]
> performance is too unpredictable to be useful with the setup I have

If you haven't already: I'd recommend to try building current master as opposed to the official release. There is a large amount of content-discovery improvement work being done for the past several months.

If that doesn't help still - it would be useful to know a bit more about your painpoints, either as a structured bugreport in this repo or a general observation over at https://discuss.ipfs.io.

In any case - please close this ticket if the current ask has been adequately addressed :)

**Comment by tmbdev** [comment]
Thanks for the response, and I'll give that a try.

I still think that better progress info would be helpful both for users and (probably) developers. That is, for an operation, it would be nice to see in real time which servers are contacted, what parts of the file are being downloaded from where, what the download speeds are, etc. Something a little bit like Etherape, though a real-time text display somewhat like "docker build" users would be fine too.
